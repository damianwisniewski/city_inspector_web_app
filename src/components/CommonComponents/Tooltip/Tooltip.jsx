import React, { Component } from 'react'
import PropTypes from 'prop-types'

import './Tooltip.scss'

class Tooltip extends Component {
	constructor(props) {
		super(props)
		this.htmlRef = React.createRef()
	}

	componentDidMount() {
		this.htmlRef.current.parentNode.classList.add('drop-down-parent')
	}

	render() {
		const { positionVertical, positionHorizontal, hoverable, show } = this.props

		return (
			<div
				ref={this.htmlRef}
				className={`drop-down drop-down--${positionVertical}-${positionHorizontal}${
					hoverable ? ' drop-down--hoverable' : ''
				}${show ? ' drop-down--show' : ''}`}
			>
				{this.props.children}
			</div>
		)
	}
}

Tooltip.defaultProps = {
	positionVertical: 'bottom',
	positionHorizontal: 'center',
	hoverable: true,
	show: false,
}

Tooltip.propTypes = {
	children: PropTypes.node.isRequired,
	positionVertical: PropTypes.oneOf(['top', 'bottom']),
	positionHorizontal: PropTypes.oneOf(['left', 'center', 'right']),
	hoverable: PropTypes.bool,
	show: PropTypes.bool,
}

export default Tooltip
