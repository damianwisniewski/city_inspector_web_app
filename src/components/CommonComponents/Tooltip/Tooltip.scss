@import '../../../assets/styleHelpers/index';

@mixin arrow-down {
	position: absolute;

	width: 0;
	height: 0;

	border-top: 15px solid $light_gray;
	border-right: 5px solid transparent;
	border-left: 5px solid transparent;

	content: '';
}

@mixin arrow-up {
	position: absolute;

	width: 0;
	height: 0;

	border-right: 5px solid transparent;
	border-bottom: 15px solid $light_gray;
	border-left: 5px solid transparent;

	content: '';
}

.drop-down-parent {
	position: relative;

	.drop-down {
		position: absolute;
		z-index: 1000;

		padding: 5px;

		font-size: 11px;

		background-color: $light_gray;
		border-radius: 5px;
		box-shadow: $shadow;
		visibility: hidden;
		opacity: 0;

		transition: opacity 0.2s linear;

		&--top {
			&-center {
				top: 0;
				left: 50%;

				transform: translate(-50%, calc(-100% - 14px));

				&::before {
					@include arrow-down();
					bottom: -15px;
					left: 50%;

					transform: translateX(-50%);
					filter: drop-shadow(0px 3px 2px rgba(0, 0, 0, 0.6));
				}
			}

			&-left {
				top: 0;
				left: 0;

				transform: translate(-90%, calc(-100% - 14px));

				&::before {
					@include arrow-down();
					right: 0;
					bottom: -15px;

					transform: translateX(-50%);
					filter: drop-shadow(0px 3px 2px rgba(0, 0, 0, 0.6));
				}
			}

			&-right {
				top: 0;
				right: 0;

				transform: translate(90%, calc(-100% - 14px));

				&::before {
					@include arrow-down();
					bottom: -15px;
					left: 0;

					transform: translateX(50%);
					filter: drop-shadow(0px 3px 2px rgba(0, 0, 0, 0.6));
				}
			}
		}

		&--bottom {
			&-center {
				bottom: 0;
				left: 50%;

				transform: translate(-50%, calc(100% + 14px));

				&::before {
					@include arrow-up();
					top: -15px;
					left: 50%;

					transform: translateX(-50%);
					filter: drop-shadow(0px -2px 1px rgba(0, 0, 0, 0.17));
				}
			}

			&-left {
				bottom: 0;
				left: 0;

				transform: translate(-90%, calc(100% + 14px));

				&::before {
					@include arrow-up();
					top: -15px;
					right: 0;

					transform: translateX(-50%);
					filter: drop-shadow(0px -2px 1px rgba(0, 0, 0, 0.17));
				}
			}

			&-right {
				right: 0;
				bottom: 0;

				transform: translate(90%, calc(100% + 14px));

				&::before {
					@include arrow-up();
					top: -15px;
					left: 0;

					transform: translateX(50%);
					filter: drop-shadow(0px -2px 1px rgba(0, 0, 0, 0.17));
				}
			}
		}

		&--show {
			visibility: visible;
			opacity: 1;
		}
	}

	&:hover .drop-down--hoverable {
		visibility: visible;
		opacity: 1;
	}
}
