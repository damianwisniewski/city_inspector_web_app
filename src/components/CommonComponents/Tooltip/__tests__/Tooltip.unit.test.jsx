import React from 'react'
import { configure, shallow, ShallowWrapper } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import Tooltip from '../Tooltip'

configure({ adapter: new Adapter() })

describe('<Tooltip />', () => {
	/**
	 * @type {ShallowWrapper}
	 */
	let StatusElement
	const FakeMessage = 'ExampleText'

	beforeEach(() => {
		StatusElement = shallow(<Tooltip>{FakeMessage}</Tooltip>, { disableLifecycleMethods: true })
	})

	test('should render correctly Tooltip', () => {
		expect(toJson(StatusElement)).toMatchSnapshot()
	})

	test('should render correctly Tooltip with attr hoverable', () => {
		StatusElement.setProps({ hoverable: true })
		expect(toJson(StatusElement)).toMatchSnapshot()

		StatusElement.setProps({ hoverable: false })
		expect(toJson(StatusElement)).toMatchSnapshot()
	})

	test('should render correctly Tooltip with attr show', () => {
		StatusElement.setProps({ show: true })
		expect(toJson(StatusElement)).toMatchSnapshot()

		StatusElement.setProps({ show: false })
		expect(toJson(StatusElement)).toMatchSnapshot()
	})

	test('should render correctly Tooltip with positionVertical=top and one any positionHorizontal', () => {
		StatusElement.setProps({ positionVertical: 'top' })

		StatusElement.setProps({ positionHorizontal: 'left' })
		expect(toJson(StatusElement)).toMatchSnapshot()

		StatusElement.setProps({ positionHorizontal: 'center' })
		expect(toJson(StatusElement)).toMatchSnapshot()

		StatusElement.setProps({ positionHorizontal: 'right' })
		expect(toJson(StatusElement)).toMatchSnapshot()
	})

	test('should render correctly Tooltip with positionVertical=bottom and one any positionHorizontal', () => {
		StatusElement.setProps({ positionVertical: 'bottom' })

		StatusElement.setProps({ positionHorizontal: 'left' })
		expect(toJson(StatusElement)).toMatchSnapshot()

		StatusElement.setProps({ positionHorizontal: 'center' })
		expect(toJson(StatusElement)).toMatchSnapshot()

		StatusElement.setProps({ positionHorizontal: 'right' })
		expect(toJson(StatusElement)).toMatchSnapshot()
	})
})
