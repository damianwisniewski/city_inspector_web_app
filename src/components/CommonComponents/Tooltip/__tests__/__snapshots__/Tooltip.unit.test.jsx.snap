// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> should render correctly Tooltip 1`] = `
<div
  className="drop-down drop-down--bottom-center drop-down--hoverable"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with attr hoverable 1`] = `
<div
  className="drop-down drop-down--bottom-center drop-down--hoverable"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with attr hoverable 2`] = `
<div
  className="drop-down drop-down--bottom-center"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with attr show 1`] = `
<div
  className="drop-down drop-down--bottom-center drop-down--hoverable drop-down--show"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with attr show 2`] = `
<div
  className="drop-down drop-down--bottom-center drop-down--hoverable"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with positionVertical=bottom and one any positionHorizontal 1`] = `
<div
  className="drop-down drop-down--bottom-left drop-down--hoverable"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with positionVertical=bottom and one any positionHorizontal 2`] = `
<div
  className="drop-down drop-down--bottom-center drop-down--hoverable"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with positionVertical=bottom and one any positionHorizontal 3`] = `
<div
  className="drop-down drop-down--bottom-right drop-down--hoverable"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with positionVertical=top and one any positionHorizontal 1`] = `
<div
  className="drop-down drop-down--top-left drop-down--hoverable"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with positionVertical=top and one any positionHorizontal 2`] = `
<div
  className="drop-down drop-down--top-center drop-down--hoverable"
>
  ExampleText
</div>
`;

exports[`<Tooltip /> should render correctly Tooltip with positionVertical=top and one any positionHorizontal 3`] = `
<div
  className="drop-down drop-down--top-right drop-down--hoverable"
>
  ExampleText
</div>
`;
